include/shuffle/PhiloxShuffle.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+#include "shuffle/BijectiveFunctionScanShuffle.h"
+
+__host__ __device__ uint32_t mulhilo(uint64_t a, uint32_t b, uint32_t &hip) {
+  uint64_t product = a * uint64_t(b);
+  hip = product >> 32;
+  return uint32_t(product);
+}
+
+template <uint64_t num_rounds = 16> class PhiloxBijectiveFunction {
+public:
+  template <class RandomGenerator>
+  void init(uint64_t capacity, RandomGenerator &random_function) {
+    uint64_t total_bits = getCipherBits(capacity);
+    // Half bits rounded down
+    left_side_bits = total_bits / 2;
+    left_side_mask = (1ull << left_side_bits) - 1;
+    // Half the bits rounded up
+    right_side_bits = total_bits - left_side_bits;
+    right_side_mask = (1ull << right_side_bits) - 1;
+    for (int i = 0; i < num_rounds; i++) {
+      key[i] = random_function();
+    }
+  }
+
+  uint64_t getMappingRange() const {
+    return 1ull << (left_side_bits + right_side_bits);
+  }
+
+  __host__ __device__ uint64_t operator()(const uint64_t val) const {
+    uint32_t state[2] = {uint32_t(val >> right_side_bits),
+                         uint32_t(val & right_side_mask)};
+    for (int i = 0; i < num_rounds; i++) {
+      uint32_t hi;
+      uint32_t lo = mulhilo(M0, state[0], hi);
+      lo = (lo << (right_side_bits - left_side_bits)) |
+           state[1] >> left_side_bits;
+      state[0] = ((hi ^ key[i]) ^ state[1]) & left_side_mask;
+      state[1] = lo & right_side_mask;
```

Comment:
Are you sure this is bijective? Im not convinced that this a) provides a bijection b) has the desired avalanche for the reduced number of bits

Reply:
On a) as far as I can tell the algorithm relies on the fact that lo is a bijection so it can be reversed with perfect knowledge to get state[0]. However that doesnt guarantee that any arbitrary subset of bits will also be a bijection, so it seems to me that it would be possible for two numbers to produce an identical `hi` and `lo` value in the bits we consider and thus an identical final result. Such a case would only occur with low probability and only at large sizes so it is unlikely the tests would catch this

Reply:
To make this work we would need to find a constant for every number of bits supported (1-48 say) and split at half the number of bits. I would have to think about how to deal with odd length keys but I think dropping the upper bits of hi should be acceptable since it is only used for diffusion

Reply:
> so it seems to me that it would be possible for two numbers to produce an identical hi and lo value in the bits we consider and thus an identical final result.

This is a valid output though right? That one of the possible bijections is the identity? The only problem is if it occurs more frequently or less frequently than it should.

Reply:
:thinking: Thinking some more it might be ok to assume lo is a bijection. If we can guarantee state[0] has fewer bits/same number of bits as lo I cant see why the claim that it is a bijection wouldn't follow. But I havent seen their proof so it may require the constant to also be that many bits or fewer

Reply:
Oh actually, it is a bijection because its equivalent to our LCG algorithm if the constant is odd.

Reply:
>This is a valid output though right? That one of the possible bijections is the identity? The only problem is if it occurs more frequently or less frequently than it should.

This wouldnt be valid as two different numbers would produce a the same result and collide

Reply:
Pretty confident its correct. Have written up the details.